AppWidget/app/src/main/java/com/example/android/appwidget/glance/image/ImageGlanceWidget.kt

```diff
@@ -0,0 +1,187 @@
+                    .fillMaxSize()
+                    .appWidgetBackground()
+                    .background(GlanceTheme.colors.background)
+                    .appWidgetBackgroundCornerRadius(),
+                contentAlignment = if (imagePath == null) {
+                    Alignment.Center
+                } else {
+                    Alignment.BottomEnd
+                }
+            ) {
+                if (imagePath != null) {
+                    Image(
+                        provider = getImageProvider(imagePath),
+                        contentDescription = null,
+                        contentScale = ContentScale.FillBounds,
+                        modifier = GlanceModifier
+                            .fillMaxSize()
+                            .clickable(actionRunCallback<RefreshAction>())
+                    )
+                    Text(
+                        text = "Source: ${currentState(sourceKey)}",
+                        style = TextStyle(
+                            color = GlanceTheme.colors.onPrimary,
+                            fontSize = 12.sp,
+                            fontStyle = FontStyle.Italic,
+                            textAlign = TextAlign.End,
+                            textDecoration = TextDecoration.Underline
+                        ),
+                        modifier = GlanceModifier
+                            .fillMaxWidth()
+                            .padding(8.dp)
+                            .background(GlanceTheme.colors.primary)
+                            .clickable(
+                                actionStartActivity(
+                                    Intent(
+                                        Intent.ACTION_VIEW,
+                                        Uri.parse(currentState(sourceUrlKey))
+                                    )
+                                )
+                            )
+                    )
+                } else {
+                    CircularProgressIndicator()
+
+                    // Enqueue the worker after the composition is completed using the glanceId as
+                    // tag so we can cancel all jobs in case the widget instance is deleted
+                    val glanceId = LocalGlanceId.current
+                    SideEffect {
+                        ImageWorker.enqueue(context, size, glanceId)
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Called when the widget instance is deleted. We can then cleanup any ongoing task.
+     */
+    override suspend fun onDelete(context: Context, glanceId: GlanceId) {
+        super.onDelete(context, glanceId)
+        ImageWorker.cancel(context, glanceId)
+    }
+
+    /**
+     * Get the bitmap from the cache file
+     *
+     * Note: since it's a single image resized to the available space we probably won't reach the
+     * memory limit, otherwise you need to generate a URI granting permissions to the launcher.
```

Comment:
Suggest:

     * Note: Because it's a single image resized to the available space, you
     * probably won't reach the memory limit. If you do reach the memory limit,
     * you'll need to generate a URI granting permissions to the launcher.